src/librustc_mir/transform/type_check.rs

```diff
@@ -1008,6 +1009,102 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
+                (&def.variants[variant], substs)
+            },
+            AggregateKind::Closure(def_id, substs) => {
+                return match substs.upvar_tys(def_id, tcx).nth(field) {
+                    Some(ty) => Ok(ty),
+                    None => Err(FieldAccessError::OutOfRange {
+                        field_count: substs.upvar_tys(def_id, tcx).count()
+                    }),
+                }
+            },
+            AggregateKind::Generator(def_id, substs, _) => {
+                if let Some(ty) = substs.upvar_tys(def_id, tcx).nth(field) {
+                    return Ok(ty);
+                }
+
+                return match substs.field_tys(def_id, tcx).nth(field) {
+                    Some(ty) => Ok(ty),
+                    None => Err(FieldAccessError::OutOfRange {
+                        field_count: substs.field_tys(def_id, tcx).count() + 1
+                    }),
+                }
+            },
+            AggregateKind::Array(ty) => {
+                return Ok(ty);
+            },
+            AggregateKind::Tuple => {
+                unreachable!("This should have been covered in check_rvalues");
+            },
+        };
+
+        if let Some(field) = variant.fields.get(field) {
+            Ok(self.normalize(&field.ty(tcx, substs), location))
+        } else {
+            Err(FieldAccessError::OutOfRange { field_count: variant.fields.len() })
+        }
+    }
+
+    fn check_rvalue(&mut self, mir: &Mir<'tcx>, rv: &Rvalue<'tcx>, location: Location) {
+        let tcx = self.tcx();
+        match rv {
+            Rvalue::Aggregate(ak, ops) => {
+                match **ak {
+                    // tuple rvalue field type is always the type of the op. Nothing to check here.
+                    AggregateKind::Tuple => { },
+                    _ => {
+                        for (i, op) in ops.iter().enumerate() {
+                            let field_ty = match self.aggregate_field_ty(ak, i, location) {
+                                Ok(field_ty) => field_ty,
+                                Err(FieldAccessError::OutOfRange { field_count }) => {
+                                    span_mirbug!(
+                                        self,
+                                        rv,
+                                        "accessed field #{} but variant only has {}",
+                                        i,
+                                        field_count);
+                                    continue;
+                                },
+                            };
+                            let op_ty = match op {
+                                Operand::Consume(lv) => lv.ty(mir, tcx).to_ty(tcx),
+                                Operand::Constant(c) => c.ty,
+                            };
+                            if let Err(terr) = self.sub_types(op_ty, field_ty, location.at_successor_within_block()) {
```

Comment:
This line causes the follow error to occur during compilation:
```
   Compiling core v0.0.0 (file:///home/paulf/rust2/src/libcore)
error: internal compiler error: broken MIR in NodeId(258966) (str::pattern::CharSearcher::{{constructor}}(_1,)): <str::pattern::CharEqPattern<char> as str::pattern::Pattern>::Searcher is not a subtype of str::pattern::CharEqSearcher<'_, _>: Sorts(ExpectedFound { expected: str::pattern::CharEqSearcher<'_, _>, found: <str::pattern::CharEqPattern<char> as str::pattern::Pattern>::Searcher })
   --> src/libcore/str/pattern.rs:418:1
    |
418 | pub struct CharSearcher<'a>(<CharEqPattern<char> as Pattern<'a>>::Searcher);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
```
Do I need to do additional work to get the types to match or is this a legitimate bug?